go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -364,3 +631,562 @@ func (vc *vcopier) fastForward(ctx context.Context, copyState map[string]*sqltyp
+	workerFactory := func(_ context.Context) (*vcopierCopyWorker, error) {
+		return newVCopierCopyWorker(
+			false, /* close db client */
+			vc.vr.dbClient,
+		), nil
+	}
+	if isExperimentalParallelBulkInsertsEnabled() {
+		workerFactory = func(ctx context.Context) (*vcopierCopyWorker, error) {
+			dbClient, err := vc.newClientConnection(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create new db client: %s", err.Error())
+			}
+			return newVCopierCopyWorker(
+				true, /* close db client */
+				dbClient,
+			), nil
+		}
+	}
+	return workerFactory
+}
+
+// Enqueue a new copy task. This will obtain a worker from the pool, execute
+// the task with that worker, and afterwards return the worker to the pool. If
+// vcopierCopyWorkQueue is configured to operate concurrently, the task will be
+// executed in a separate goroutine. Otherwise the task will be executed in the
+// calling thread.
+func (vcq *vcopierCopyWorkQueue) Enqueue(ctx context.Context, currT *vcopierCopyTask) error {
+	if !vcq.IsOpen() {
+		return fmt.Errorf("work queue is not open")
+	}
+
+	// Get a handle on an unused worker.
+	poolH, err := vcq.workerPool.Get(ctx, []string{})
+	if err != nil {
+		return fmt.Errorf("failed to get a worker from pool: %s", err.Error())
+	}
+
+	currW := poolH.(*vcopierCopyWorker)
+
+	execute := func(task *vcopierCopyTask) {
+		currW.Execute(ctx, task)
+		vcq.workerPool.Put(poolH)
+	}
+
+	// If the work queue is configured to work concurrently, execute the task
+	// in a separate goroutine. Otherwise execute the task in the calling
+	// goroutine.
+	if vcq.concurrent {
+		go execute(currT)
+	} else {
+		execute(currT)
+	}
+
+	return nil
+}
+
+// Close waits for all workers to be returned to the worker pool.
+func (vcq *vcopierCopyWorkQueue) Close() {
+	if !vcq.IsOpen() {
+		return
+	}
+	vcq.isOpen = false
+	vcq.workerPool.Close()
+}
+
+func (vcq *vcopierCopyWorkQueue) IsOpen() bool {
+	return vcq.isOpen
+}
+
+// Open the work queue. The provided arguments are used to generate INSERT and
+// UPDATE statements.
+func (vcq *vcopierCopyWorkQueue) Open(
+	copyStateUpdate *sqlparser.ParsedQuery,
+	pkfields []*querypb.Field,
+	tablePlan *TablePlan,
+) {
+	if vcq.IsOpen() {
+		return
+	}
+
+	poolCapacity := int(math.Max(float64(vcq.maxDepth), 1))
+	vcq.workerPool = pools.NewResourcePool(
+		/* factory */
+		func(ctx context.Context) (pools.Resource, error) {
+			worker, err := vcq.workerFactory(ctx)
+			if err != nil {
+				return nil, fmt.Errorf(
+					"failed to create copier for resource pool: %s",
+					err.Error(),
+				)
+			}
+			worker.Open(copyStateUpdate, pkfields, tablePlan)
+			return worker, nil
+		},
+		poolCapacity, /* initial capacity */
+		poolCapacity, /* max capacity */
+		0,            /* idle timeout */
+		nil,          /* log wait */
+		nil,          /* refresh check */
+		0,            /* refresh interval */
+	)
+
+	vcq.isOpen = true
+}
+
+func (vct *vcopierCopyTask) Args() *vcopierCopyTaskArgs {
```

Comment:
Do you think we'll ever need/want these to be accessible outside of the package? It feels like implementation details to me, but maybe not? I actually see the exported functions used (e.g. `vct.IsOpen()`) within the package, are there any reasons for that? Maybe something I'm not aware of.

Reply:
I think for all of these new types, we don't need to be exporting anything. Also based on your questions I think there's a Golang convention I wasn't aware of until now (a package shouldn't use functions that it exports).

Since none of the methods on these new types need to be exported, I'll make them private.